### PR TITLE
add the filebrowser rock and tests

### LINKED ISSUE #72: update the filebrowser rock for Charmed Kubeflow 1.8
### Context

The filebrowser/filebrowser:latest image [was introduced](https://github.com/kubeflow/manifests/pull/2522/files#diff-1e24d4b884def12df3f5566af49503051c1e9ff7d01b9718d8828eefad7439d9R23-R24) in a recent release.  Upstream usage is [here](https://github.com/TobiasGoerke/kubeflow/blob/309abc07e213f8371ff945e848a13467a9ef1218/components/crud-web-apps/volumes/manifests/base/deployment.yaml#L24).  We need to create and use a rock for this image.

### What needs to get done

1. the rockcraft.yaml is created and builds successfully (possibly under a different repository - that is tbd)

### Definition of Done

1. the rock is published to dockerhub

### PR DESCRIPTION
**Pull request by ca-scribner**
This PR adds:

* `filebrowser/rockcraft.yaml`, which implements a rock for [filebrowser](https://github.com/filebrowser/filebrowser)
* `filebrowser/README.md` with some notes and instructions for future maintenance
* `filebrowser/tox.ini` and related sanity tests

Closes https://github.com/canonical/kubeflow-rocks/issues/72

## Testing instructions:

testing the rock itself:

```
cd filebrowser
tox -e pack
tox -e export-to-docker
docker run -p 8080:80 -e FB_ADDRESS=0.0.0.0 filebrowser:v2.27.0
# Browse in a browser to `localhost:8080` and see the app
```

(if you want to) testing in charmed kubeflow

```
# with a fully deployed charmed kubeflow
juju config kubeflow-volumes volume-viewer-image=local/filebrowser:v2.27.0

tox -e export-to-docker
docker tag filebrowser:v2.27.0 local/filebrowser:v2.27.0  # so we really know its the right image
docker save local/filebrowser:v2.27.0 > local_filebrowser_v2.27.0.tar
microk8s ctr image import local_filebrowser_v2.27.0.tar
```

then
* log into 10.64.140.43.nip.io using user123/user123
* go to volumes web app
* create a volume named testvol
* click the "Open PVC Viewer" button (small folder icon on the right side of the row)
* filebrowser window should open (although there will be no files since pvc is empty - you can create some though!)
* confirm we got a filebrowser with the new image:
```
kubectl -n user123 get pod $(kubectl get pod -n user123 --selector=app.kubernetes.io/instance=pvcviewer-testvol --output=jsonpath={.items..metadata.name}) -o yaml | grep filebrowser:
# Should return:
filebrowser:
    image: docker.io/library/local/filebrowser:v2.27.0
    image: docker.io/library/local/filebrowser:v2.27.0

# NOT
filebrowser:
    image: docker.io/library/filebrowser/filebrowser:v2.23.0
    image: docker.io/library/filebrowser/filebrowser:v2.23.0
```

### CODE REVIEW

**Comment by ca-scribner** [comment]
Hit a lot of issues while creating this rock.  They're documented in the rock and included `README.md`, but in summary the rock:
* uses a few atypical build procedures (GoReleaser, and also its `go build` embeds the frontend into the application)
* a healthcheck script and `HEALTHCHECK` directive in the dockerfile
* needs somewhere with write access for a local db
* needs a config file to define defaults to both the healthcheck and filebrowser binary
* Charmed Kubeflow uses it via a controller we don't control, so we need to emulate the API of this image to make it a drop-in replacement of the upstream docker image.  This means putting files in consistent places and setting a default `entrypoint-service`

Along the way I also got stuck with rockcraft in a few places:
* [misunderstanding how to pass args](https://github.com/canonical/rockcraft/issues/476) for the `entrypoint-service`
* finding out that rocks [don't automatically stage `.dotfiles`](https://matrix.to/#/!NPPCseDHKRvSBMUEXN:ubuntu.com/$3pkoT_gtSQlIa_J2lv6iy3P0JdzMQ5Cbt38guXG9Sg4?via=ubuntu.com&via=matrix.org)
* [finding a pattern for passing build artifacts between parts](https://matrix.to/#/!NPPCseDHKRvSBMUEXN:ubuntu.com/$NGm6PTH62PVOjt2GtAAxzaGr_3BGap1COXcZUwx1voI?via=ubuntu.com&via=matrix.org)
* [finding that `stage` clobbers `stage-packages`](https://matrix.to/#/!NPPCseDHKRvSBMUEXN:ubuntu.com/$8nyOhyDbS-NqZOB9TIdwqezGsI2ibQYsEr_oxIX8t6k?via=ubuntu.com&via=matrix.org), meaning that if you set `stage` you no longer get your `stage-packages`
* [figuring out how `permissions` work](https://matrix.to/#/!NPPCseDHKRvSBMUEXN:ubuntu.com/$4XxxiLC2Vsn2-NAIqZ-Z0w_SJJzOQXTYb7cBmnTUStU?via=ubuntu.com&via=matrix.org)
* [clarifying how different directories work in rockcraft](https://matrix.to/#/!NPPCseDHKRvSBMUEXN:ubuntu.com/$5HDSvE0-TKVy44uWnNuBqd2LNYlLqwBTvhMxo7j4nA8?via=ubuntu.com&via=matrix.org)

File: filebrowser/README.md
Comment (orfeas-k) [inline filebrowser/README.md]: Should those be moved to CONTRIBUTING.md? The only reason I 'm skeptical about this because we don't have a standard practice yet around this.

File: filebrowser/README.md
Comment (orfeas-k) [inline filebrowser/README.md:1]: Links could refer to a specific upstream version instead of main.

File: filebrowser/rockcraft.yaml
Comment (orfeas-k) [inline filebrowser/rockcraft.yaml:1]: Regarding repo's structure, do we need a second `filebrowser` directory? This looks like it is a single rock repository

File: filebrowser/README.md
Comment (orfeas-k) [inline filebrowser/README.md]: props for the detailed explanation though. It helps understand the rock and review with more context.

File: filebrowser/rockcraft.yaml
Comment (orfeas-k) [inline filebrowser/rockcraft.yaml]: Is this needed though? If not, should the upstream convention be a reason to diverge from our standard practice?

File: filebrowser/rockcraft.yaml
Comment (orfeas-k) [inline filebrowser/rockcraft.yaml]: I commented that [here](https://github.com/canonical/kubeflow-rocks/pull/74#pullrequestreview-1872087477) as well where we agreed with Daniela to keep it as `replace` and discuss in the next `rocks` meeting. I 'm fine with keeping it as is or using `replace`, since I don't think it makes any difference in this case. I 'll leave this up to you.

File: filebrowser/rockcraft.yaml
Comment (orfeas-k) [inline filebrowser/rockcraft.yaml:22]: Good to see you figured out how this works for all of us.

File: filebrowser/rockcraft.yaml
Comment (orfeas-k) [inline filebrowser/rockcraft.yaml:33]: should we consider adding a timeout? I see you used `period` from upstream definition ` --start-period=2s --interval=5s --timeout=3s`

Comment (orfeas-k) [reply filebrowser/rockcraft.yaml:33]: or even sth for start-period?

File: filebrowser/rockcraft.yaml
Comment (orfeas-k) [inline filebrowser/rockcraft.yaml]:
```suggestion
  create-srv-dir:
```
Do I miss sth?

File: filebrowser/rockcraft.yaml
Comment (orfeas-k) [inline filebrowser/rockcraft.yaml]:
```suggestion
    # Note: If we write files to this dir as part of the rock build, create-srv-dir
```

File: filebrowser/rockcraft.yaml
Comment (orfeas-k) [inline filebrowser/rockcraft.yaml:145]: I was thinking of why `install` part instead of `prime`. But then I saw this is under the questions we have for rockcraft team. (you can resolve this)

File: filebrowser/rockcraft.yaml
Comment (orfeas-k) [inline filebrowser/rockcraft.yaml:62]: Could `healthcheck` and `create-srv-directory` be into one with a name of `copy-files` or `structure-final-image`? They look like doing exactly the same thing. That being said, if you 'd like to keep them separate, for having each part doing exactly one thing, it's fine by me. I 'm not sure if there's a stylistic guide on if it's better that each part does one thing exactly.

File: filebrowser/rockcraft.yaml
Comment (orfeas-k) [inline filebrowser/rockcraft.yaml:71]: Good job on figuring this out too. Do we know that we don't need any of the other fields of [docker_config.json](https://github.com/filebrowser/filebrowser/blob/d49c3dfacfc0ff07e620b3ad2700e64927b06235/docker_config.json)?

File: filebrowser/rockcraft.yaml
Comment (orfeas-k) [inline filebrowser/rockcraft.yaml]: **nit:** That's a good comment to have for future reference. But do you think we need the last 3 lines as well?

File: filebrowser/rockcraft.yaml
Comment (orfeas-k) [inline filebrowser/rockcraft.yaml]: Can we install npm as a `build-package`? I found this [example](https://github.com/canonical/traefik-rock/blob/9e9ded585dc16d1f27374b79f2d20e91d883dd4e/2.10.6/rockcraft.yaml#L31-L33) (or even [snap](https://github.com/canonical/prometheus-rock/blob/70c0ca78898df2849ffd7f039a19cf1b6dddcba9/2.46.0/rockcraft.yaml#L21-L23))

File: filebrowser/rockcraft.yaml
Comment (orfeas-k) [inline filebrowser/rockcraft.yaml]: **nit:** why break the style of using `-` for matrix in the whole rockcraft.yaml file? (sorry :grin:)

File: filebrowser/rockcraft.yaml
Comment (orfeas-k) [inline filebrowser/rockcraft.yaml:104]: :exploding_head:

File: filebrowser/tox.ini
Comment (orfeas-k) [inline filebrowser/tox.ini]:
```suggestion
             DOCKER_IMAGE=$NAME:$VERSION && \
```
I think this was copied from somewhere where this is wrong as well (I remember introducing that error somewhere :grimacing: )

**Review by orfeas-k — COMMENTED** [review]
Great job @ca-scribner . This had quite a few hard spots that needed to go around. I left some comments

File: filebrowser/README.md
Comment (ca-scribner) [reply filebrowser/README.md]: lgtm!  done

File: filebrowser/README.md
Comment (ca-scribner) [reply filebrowser/README.md:1]: Yeah I wondered about this too - should the generic "update by doing X" guide cite specific versions?  I wasn't sure.  Definitely the rockcraft.yaml should be clear what version it follows.  

I added a specific link to the rockcraft.yaml header - should I modify contributing.md as well?

File: filebrowser/rockcraft.yaml
Comment (ca-scribner) [reply filebrowser/rockcraft.yaml:1]: Great question.  It does feel redundant.  Do you know if we have any single rock repos?  I worried about breaking CI or other conventions by making a single rock repo

I think [get-rocks-modified](https://github.com/canonical/charmed-kubeflow-workflows/blob/main/.github/workflows/get-rocks-modified.yaml) would be ok with this, but not sure if we made the assumption anywhere else

File: filebrowser/rockcraft.yaml
Comment (ca-scribner) [reply filebrowser/rockcraft.yaml]: Yeah I know what you mean.  Adding this to our daily to decide tomorrow.

File: filebrowser/rockcraft.yaml
Comment (ca-scribner) [reply filebrowser/rockcraft.yaml]: I think I just copied whatever something else used :D

I'll switch to `replace` for now as it feels like it makes more sense, and we can update differently in future

File: filebrowser/rockcraft.yaml
Comment (ca-scribner) [reply filebrowser/rockcraft.yaml:33]: Added `timeout: 3s`.  I don't see a `start-period` equivalent [in pebble](https://github.com/canonical/pebble?tab=readme-ov-file#layer-specification)

File: filebrowser/rockcraft.yaml
Comment (ca-scribner) [reply filebrowser/rockcraft.yaml:62]: definitely `healthcheck` and `add-config` could be combined, and probably also `create-srv-directory`.  To me I like them separate as it feels easier to debug, but I'm flexible.  

I don't know if there's a stylistic guide either.

File: filebrowser/rockcraft.yaml
Comment (ca-scribner) [reply filebrowser/rockcraft.yaml:71]: Its a bit subtle but this **edits** the `.database` field and **preserves** all the other fields.  So the end result is a `.filebrowser.json` file of:

```json
{
  "port": 80,
  "baseURL": "",
  "address": "",
  "log": "stdout",
  "database": "/tmp/database.db",
  "root": "/srv"
}
```

File: filebrowser/rockcraft.yaml
Comment (ca-scribner) [reply filebrowser/rockcraft.yaml]: Yeah that's fair.  I will take them out

File: filebrowser/rockcraft.yaml
Comment (ca-scribner) [reply filebrowser/rockcraft.yaml]: Great question - hadn't spotted that!  Probably we don't care about the minor version so snap sounds helpful.  Let me try it

File: filebrowser/rockcraft.yaml
Comment (ca-scribner) [reply filebrowser/rockcraft.yaml:104]: don't get me started, @orfeas-k :laughing:

File: filebrowser/rockcraft.yaml
Comment (ca-scribner) [reply filebrowser/rockcraft.yaml]: Nope, it was I who missed something :)  fixed

File: filebrowser/rockcraft.yaml
Comment (ca-scribner) [reply filebrowser/rockcraft.yaml]: fixed

File: filebrowser/rockcraft.yaml
Comment (ca-scribner) [reply filebrowser/rockcraft.yaml]: :D Very fair.  Fixed

File: filebrowser/rockcraft.yaml
Comment (ca-scribner) [reply filebrowser/rockcraft.yaml:145]: yeah I think the philosophy is `build` puts its artifacts that should go forward into `install`, and then later someone might rename them, move them on to stage, then on to prime, etc etc

File: filebrowser/rockcraft.yaml
Comment (ca-scribner) [reply filebrowser/rockcraft.yaml:104]: I wonder if, for things like this that we'll likely repeat but don't want to forget it was intentional, we should have some github issues or something else we could link.  So instead of this explanation, it could just be:
```
# Override prime per http://github.com/canonical/something/issues/1234
prime: 
  - "-*"
```

it would also give us something easy to search for if we ever wanted to find and replace all our patches like this

Comment (ca-scribner) [reply filebrowser/rockcraft.yaml:104]: wdyt?  could also be linking to a style guide spec with a linkable bookmark on each item, although those links get LONG :D

File: filebrowser/rockcraft.yaml
Comment (ca-scribner) [reply filebrowser/rockcraft.yaml]: Oh that's excellent, it worked great.  ty!  changed it to a snap (wasn't sure how to pin a version for the `build-packages`)

File: filebrowser/README.md
Comment (orfeas-k) [reply filebrowser/README.md:1]: Great. I think not having to update the `CONTRIBUTING.md` file links each time we update the version is fine, as long as there are no changes introduced. I 'm fine as this is now (pointing to a specific commit) which makes sure that it won't point to something false in the future. You can resolve once you see this.

File: filebrowser/rockcraft.yaml
Comment (orfeas-k) [reply filebrowser/rockcraft.yaml:1]: We have this [metacontroller-rock](https://github.com/canonical/metacontroller-rock). In [OIDC](https://github.com/canonical/oidc-authservice-rock) though, we have followed what is done here. I 'm in favor of removing the redundant directory. Our CI should be able to handle this AFAICT. However, if not, I don't think it would be hard extending for something like that, since it's a limitation.

File: filebrowser/rockcraft.yaml
Comment (orfeas-k) [reply filebrowser/rockcraft.yaml:33]: Hmm, I thought I had seen sth somewhere that introduced a delay before starting the service or doing health checks implemented by Phoevos, but I don't see anything in the layer specification too. It's fine as is then, you can resolve!

File: filebrowser/rockcraft.yaml
Comment (orfeas-k) [reply filebrowser/rockcraft.yaml:62]: I 'm really torn on this cause I just see pros and cons on both sides. I 'll say let's keep it that way then, since we haven't standardized this. Should we add a note about style guide in the ROCKs meeting?

File: filebrowser/rockcraft.yaml
Comment (orfeas-k) [reply filebrowser/rockcraft.yaml:71]: pff `jq` and `yq` always get me confused. Looks good then

File: filebrowser/rockcraft.yaml
Comment (orfeas-k) [reply filebrowser/rockcraft.yaml:104]: Good points, both of these. I think having a style guide is needed more and more as we create more ROCKs, but maybe until we actually create one, we can have issues (which could also be closed) that document stylistic decisions or behaviours we need to implement in order patch rockcraft default behaviour. On the long links point, you 're right but it's not a blocker I guess.

File: filebrowser/tox.ini
Comment (orfeas-k) [inline filebrowser/tox.ini]:
```suggestion
             echo "Exporting $ROCK to docker as $DOCKER_IMAGE" && \
```
one more

File: filebrowser/rockcraft.yaml
Comment (ca-scribner) [reply filebrowser/rockcraft.yaml:1]: added this to the daily just for a 2 sec discussion, then I'll go with whatever is mentioned there

File: filebrowser/rockcraft.yaml
Comment (ca-scribner) [reply filebrowser/rockcraft.yaml:62]: +1

Comment (ca-scribner) [reply filebrowser/rockcraft.yaml:62]: I'm resolving this for now, but added a AI to the rock meeting

File: filebrowser/rockcraft.yaml
Comment (ca-scribner) [reply filebrowser/rockcraft.yaml:104]: k ill resolve this for now, but add a note to our rockcraft meeting

File: filebrowser/rockcraft.yaml
Comment (ca-scribner) [reply filebrowser/rockcraft.yaml]: Removing the v - we don't like v's!

File: filebrowser/rockcraft.yaml
Comment (ca-scribner) [inline filebrowser/rockcraft.yaml]:
```suggestion
version: 2.27.0
```

File: filebrowser/rockcraft.yaml
Comment (ca-scribner) [reply filebrowser/rockcraft.yaml:1]: resolving as we decided to stick with the nested repo

**Review by orfeas-k — APPROVED** [review]
Thank you Andrew, very good job!

File: filebrowser/rockcraft.yaml
Comment (ca-scribner) [inline filebrowser/rockcraft.yaml:89]: it has been raised in https://github.com/canonical/craft-parts/issues/427

**Pull request merged**
